master/cache.lua

```diff
@@ -151,50 +187,133 @@ local function fetch_and_cache_state_marathon()
+        ngx.log(ngx.NOTICE, "Marathon leader request failed: " .. err)
+        if not cache_data("marathonleader", nil) then
+            ngx.log(ngx.ERR, "Invalidating Marathon leader cache failed")
+        end
+        return
     end
+
+    -- https://mesosphere.github.io/marathon/docs/rest-api.html#get-v2-leader
+    if mleaderRes.status == 404 then
+        ngx.log(ngx.NOTICE, "Storing empty Marathon leader to SHM")
+        local empty_leader_json = '{"port": 0, "address": "not elected"}'
+        if not cache_data("marathonleader", empty_leader_json) then
+            ngx.log(ngx.ERR, "Storing Marathon leader cache failed")
+        end
+        return
+    end
+
+    local mleader, err = cjson_safe.decode(mleaderRes.body)
+    if not mleader then
+        ngx.log(ngx.NOTICE, "Cannot decode Marathon leader JSON: " .. err)
+        if not cache_data("marathonleader", nil) then
+            ngx.log(ngx.ERR, "Invalidating Marathon leader cache failed")
+        end
+        return
+    end
+
+    local split_mleader = mleader['leader']:split(":")
+    local parsed_mleader = {}
+    parsed_mleader["address"] = split_mleader[1]
+    parsed_mleader["port"] = split_mleader[2]
+    local mleader = cjson_safe.encode(parsed_mleader)
+
+    ngx.log(ngx.DEBUG, "Storing Marathon leader to SHM")
+    if not cache_data("marathonleader", mleader) then
+        ngx.log(ngx.ERR, "Storing Marathon leader cache failed")
+        return
+    end
+
+    ngx.update_time()
+    local time_now = ngx.now()
+    if cache_data("marathonleader_last_refresh", time_now) then
+        ngx.log(ngx.INFO, "Marathon leader cache has been successfully updated")
+    end
+
+    return
 end
 
 
-local function fetch_and_cache_state_mesos()
+local function fetch_and_store_state_mesos()
     -- Fetch state JSON summary from Mesos. If successful, store to SHM cache.
     -- Expected to run within lock context.
-    local mesosRes, err = request("leader.mesos", 5050, "/master/state-summary")
+    local mesosRes, err = request("leader.mesos", 5050, "/master/state-summary", false)
 
     if err then
         ngx.log(ngx.NOTICE, "Mesos state request failed: " .. err)
+        if not cache_data("mesosstate", nil) then
+            ngx.log(ngx.ERR, "Invalidating Mesos state cache failed")
+        end
         return
     end
 
-    ngx.update_time()
-    local time_fetched = ngx.now()
     ngx.log(ngx.DEBUG, "Storing Mesos state to SHM.")
+    if not cache_data("mesosstate", mesosRes.body) then
+        ngx.log(ngx.ERR, "Storing mesos state cache failed")
+        return
+    end
 
-    local success = cache_data("statejson", mesosRes.body)
-    if success then
-        cache_data("last_fetch_time_mesos", time_fetched)
+    ngx.update_time()
+    local time_now = ngx.now()
+    if cache_data("mesosstate_last_refresh", time_now) then
+        ngx.log(ngx.INFO, "Mesos state cache has been successfully updated")
+    end
+
+    return
+end
+
+
+local function refresh_needed(ts_name)
```

Comment:
time-stamp name. I am not attached to this name, if you have something better in mind then go ahead.